component/kafka/component.go

```diff
@@ -0,0 +1,527 @@
+	// and thus prevent the Consume loop from picking up the message again
+	if c.failStrategy == ExitStrategy {
+		c.cancel()
+	}
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (c *consumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for {
+		select {
+		case msg, ok := <-claim.Messages():
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, topic = %s", string(msg.Value), msg.Timestamp, msg.Topic)
+				topicPartitionOffsetDiffGaugeSet(c.group, msg.Topic, msg.Partition, claim.HighWaterMarkOffset(), msg.Offset)
+				messageStatusCountInc(messageReceived, c.group, msg.Topic)
+				err := c.insertMessage(session, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			c.mu.Lock()
+			err := c.flush(session)
+			if err != nil {
+				return err
+			}
+			c.mu.Unlock()
+		case <-c.ctx.Done():
+			if c.ctx.Err() != context.Canceled {
+				log.Infof("closing consumer: %v", c.ctx.Err())
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) flush(session sarama.ConsumerGroupSession) error {
+	var err error
+
+	if len(c.msgBuf) > 0 {
+		ctxCh, cancel := c.getContextWithCorrelation()
+		defer cancel()
+
+		for i := 0; i <= c.retries; i++ {
+			wrappedMessages := make([]MessageWrapper, 0, len(c.msgBuf))
+			for _, msg := range c.msgBuf {
+				messageStatusCountInc(messageProcessed, c.group, msg.Topic)
+				wrappedMessages = append(wrappedMessages, &messageWrapper{msg: msg})
+			}
+
+			err = c.proc(ctxCh, wrappedMessages)
+			if err == nil {
+				break
+			}
+
+			if c.ctx.Err() == context.Canceled {
+				return fmt.Errorf("context was cancelled after processing error: %w", err)
+			}
+
+			consumerErrorsInc(c.name)
+			if c.retries > 0 {
+				log.Errorf("failed run, retry %d/%d with %v wait: %v", i, c.retries, c.retryWait, err)
+				time.Sleep(c.retryWait)
+			}
+		}
+
+		if err != nil {
+			for _, msg := range c.msgBuf {
+				messageStatusCountInc(messageErrored, c.group, msg.Topic)
+			}
+			switch c.failStrategy {
+			case ExitStrategy:
+				log.Errorf("exhausted all retries but could not process message(s)")
+				c.err = err
+				return err
+			case SkipStrategy:
+				log.Errorf("exhausted all retries but could not process message(s) so skipping")
+			}
+		}
+
+		for _, m := range c.msgBuf {
+			session.MarkMessage(m, "")
+		}
+
+		if c.commitSync {
+			session.Commit()
+		}
+
+		c.msgBuf = make([]*sarama.ConsumerMessage, 0, c.batchSize)
+	}
+
+	return err
+}
+
+func (c *consumerHandler) getContextWithCorrelation() (context.Context, context.CancelFunc) {
+	if len(c.msgBuf) == 1 {
+		msg := c.msgBuf[0]
+		corID := getCorrelationID(msg.Headers)
+
+		_, ctxCh := trace.ConsumerSpan(c.ctx, trace.ComponentOpName(consumerComponent, msg.Topic),
```

Comment:
Notes for reviewer: no tracing for messages has been applied here. The `ConsumerSpan` is only used to generate a context in case there is one message. We need to reintroduce trace but what do you think is the most optimal way when there is batching?

Reply:
Check the SQS package which handles the span inside of ACK and NACK. Nack in kafka probably only means closing the span with error.